File: Plugins/Decoders/SafeLinksDecoder.py
from Plugins import Plugin

import urllib.parse

class SafeLinksDecoder(Plugin.Plugin):
  def __init__(self, url: str = None, name: str = "Safe Links Decoder"):
    super().__init__(name)
    self._url = url.strip() if url is not None else None

  def run(self):
    print("\n --------------------------------- ")
    print(" S A F E L I N K S   D E C O D E R  ")
    print(" --------------------------------- ")
    if self._url == None:
      self._url = str(input('Enter URL: ').strip())
    dcUrl = urllib.parse.unquote(self._url)
    dcUrl = dcUrl.replace('https://nam02.safelinks.protection.outlook.com/?url=', '')
    print(dcUrl)

File: src/plugins/decoders/safelinks_decoder.rs
use super::Plugin;
use serde_json;
use urlencoding::decode;

use crate::utils::get_input;

pub struct SafeLinksDecode {}

impl Plugin for SafeLinksDecode {
    const NAME: &'static str = "SafeLinks Decoder";

    fn new() -> Self {
        Self {}
    }

    fn perform_lookup(&self, value: &str) -> Result<serde_json::Value, String> {
        let decoded_value = Self::decode_url(value);
        Ok(serde_json::json!({ "result": "decoded", "value": decoded_value}))
    }

    fn run(&self) {
        println!("SafeLinks Decoder");

        let usr_in = get_input("Enter URL: ");
        match self.perform_lookup(&usr_in) {
          Ok(result) => println!("\n{}", result),
          Err(error) => eprintln!("Error: {}", error)
        }
    }
}

impl SafeLinksDecode {
  fn decode_url(url: &str) -> String {
    let mut dc_url = decode(url).expect("UTF-8").to_string();
    dc_url = dc_url.replace("https://nam02.safelinks.protection.outlook.com/?url=", "");
    dc_url
  }
}